thrust/zip_function.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <thrust/type_traits/integer_sequence.h>
+#include <thrust/detail/type_deduction.h>
+
+THRUST_BEGIN_NS
+
+namespace zip_detail {
+
+template <typename Function, typename Tuple, std::size_t... Is>
+__host__ __device__
+decltype(auto) apply_impl(Function&& func, Tuple&& args, index_sequence<Is...>)
+{
+  return func(thrust::get<Is>(THRUST_FWD(args))...);
+}
+
+template <typename Function, typename Tuple>
+__host__ __device__
+decltype(auto) apply(Function&& func, Tuple&& args)
+{
+  constexpr auto tuple_size = thrust::tuple_size<typename std::decay<Tuple>::type>::value;
+  return apply_impl(THRUST_FWD(func), THRUST_FWD(args), make_index_sequence<tuple_size>{});
+}
+}
+
+// Adaptor type that turns an N-ary function object into one that takes a tuple
+// of size N so it can easily be used with algorithms taking zip iterators
+template <typename Function>
+class zip_function
```

Comment:
Is `zip_function` the best name for this functionality? What is it called in other libraries?

Reply:
zip_function_adaptor? I feel like this concept is probably present in functional programming, not sure what it's called there

Reply:
I was mostly interested to know whether e.g. Boost had something similar, and if so, what they had named it. I took a quick look at Boost, but didn't see anything obviously similar. I'm fine with calling it `zip_function`.